test/src/test/java/org/corfudb/infrastructure/logreplication/LogReplicationUtilsTest.java

```diff
@@ -100,75 +68,60 @@ public void testSubscribeSnapshotSyncComplete() {
      * 4) Verify that full sync finished and snapshot sync was not considered ongoing for test_client
      * 5) Verify that full sync was not attempted and snapshot sync was ongoing for new_client
      */
-    private void testMultipleClients(boolean subscribe) {
+    @Test
+    public void testMultipleClients() throws Exception{
 
         // Snapshot sync is not in progress on test_client
         TestUtils.setSnapshotSyncOngoing(corfuStore, replicationStatusTable, client, false);
 
         // Create the listener for new_client and set snapshot sync to be in progress
         String newClient = "new_client";
-        LogReplicationListener newListener = new LogReplicationTestListener(corfuStore, namespace, newClient);
+        CountDownLatch newClientFullSyncPerformed = new CountDownLatch(1);
+        LogReplicationTestListener newListener = new LogReplicationTestListener(corfuStore,
+                namespace, newClient, newClientFullSyncPerformed);
         TestUtils.setSnapshotSyncOngoing(corfuStore, replicationStatusTable, newClient, true);
 
-        if (subscribe) {
-            String streamTag = "test_tag";
-            LogReplicationUtils.subscribe(lrListener, namespace, streamTag, new ArrayList<>(), 5, corfuStore);
-            LogReplicationUtils.subscribe(newListener, namespace, streamTag, new ArrayList<>(), 5, corfuStore);
-        } else {
-            LogReplicationUtils.attemptClientFullSync(corfuStore, lrListener, namespace);
-            LogReplicationUtils.attemptClientFullSync(corfuStore, newListener, namespace);
-        }
+        LogReplicationUtils.subscribe(lrListener, namespace, streamTag, new ArrayList<>(), 5, corfuStore);
+        LogReplicationUtils.subscribe(newListener, namespace, streamTag, new ArrayList<>(), 5, corfuStore);
 
         // Verify that full sync finished and snapshot sync was not considered ongoing for test_client
-        verifyListenerFlags((LogReplicationTestListener)lrListener, false);
+        verifyListenerFlags(lrListener, false);
 
         // Verify that full sync was not attempted and snapshot sync was ongoing for new_client
-        verifyListenerFlags((LogReplicationTestListener)newListener, true);
+        verifyListenerFlags(newListener, true);
     }
 
-    private void verifyListenerFlags(LogReplicationTestListener listener, boolean snapshotSyncOngoing) {
+    private void verifyListenerFlags(LogReplicationTestListener listener, boolean snapshotSyncOngoing) throws InterruptedException {
         if (snapshotSyncOngoing) {
-            Assert.assertTrue(listener.getClientFullSyncPending().get());
-            Assert.assertTrue(listener.getSnapshotSyncInProgress().get());
-            Assert.assertEquals(Address.NON_ADDRESS, listener.getClientFullSyncTimestamp().get());
-            Assert.assertFalse(listener.performFullSyncInvoked);
+            // We need to insert delay here to let listener workflow run before validating the latch.
+            Thread.sleep(1000);
```

Comment:
As soon as the listener gets created, the countdown latch gets set.  No wait required, right?

We have some limitations with the current tests now that the implementation has changed.  Can you add a TODO on top of this class and LogReplicationListenerIT that they need refactoring to test the current behavior properly?

Reply:
Yes that is right, but then we would just be checking the default value of the countdown latch and if for some performFullSyncAndMerge() had ran we would miss that. Also I can definitely add a todo for that.